sdk/core/Azure.Core/src/Internal/ArrayBackedPropertyBag.cs

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// A property bag which is optimized for storage of a small number of items.
+    /// If the item count is less than 2, there are no allocations. Any additional items are stored in an array which will grow as needed.
+    /// </summary>
+    internal class ArrayBackedPropertyBag<TKey, TValue> where TKey : IEquatable<TKey>
+    {
+        private int _totalCount;
+        private int _arrayCount;
+        private (TKey Key, TValue Value)[]? _array;
+        private TKey? _key1, _key2;
+        private TValue? _value1, _value2;
+
+        /// <summary>
+        /// Adds a value to the property bag.
+        /// </summary>
+        /// <param name="key">The key of the value to add.</param>
+        /// <param name="value">The value to add.</param>
+        public void Add(TKey key, TValue value)
+        {
+            Argument.AssertNotNull(key, nameof(key));
+
+            bool duplicateKeyDetected = false;
+            switch (_totalCount)
+            {
+                case 0:
+                    _key1 = key;
+                    _value1 = value;
+                    break;
+
+                case 1:
+                    if (EqualityComparer<TKey>.Default.Equals(key, _key1!))
```

Comment:
It might be worth measuring how much faster it would be to make the key long (not generic) and so use the simple == long comparison

Reply:
If we call directly `key.Equals(_key1!)`, it should we one `callvirt` difference.